A45-retry-stats.md

```diff
@@ -0,0 +1,126 @@
+A55: Exposing OpenCensus Metrics and Tracing for gRPC retry
+----
+* Author(s): dapengzhang0, dfawley, ejona86, markdroth
+* Approver: markdroth
+* Status: In Review
+* Implemented in: C-core, Java
+* Last updated: 2021-08-02
+* Discussion at: https://groups.google.com/g/grpc-io/c/EAbFqaJWp5w
+
+## Abstract
+
+This document is a proposal on how gRPC client would expose OpenCensus metrics and tracing when retry is enbaled. It can be considered as an augment of the [retry gRFC](https://github.com/grpc/proposal/blob/master/A6-client-retries.md).
+
+## Background
+
+There are a collection of OpenCensus [metrics](https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/gRPC.md#measures) for gRPC. When adding retry support to gRPC, we need to clearly define the measurement points for those metrics and add some retry specific metrics to OpenCensus library to present a clearer picture of retry attempts. The retry gRFC discussed exposing [three additional per-method metrics]((https://github.com/grpc/proposal/blob/master/A6-client-retries.md#retry-and-hedging-statistics)) for retry. However, no detail on how to record and expose these metrics was discussed. Also, these metrics still provide very limited information and the histogram view (the third metric) is not very useful because `maxAttempts` is always less than or equal 5. Besides the number of retry attempts, users may also be interested in the total number of transparent retries and the total amount of delay time caused by retry backoff.
+
+### Related Proposals: 
+* [gRPC Retry Design](https://github.com/grpc/proposal/blob/master/A6-client-retries.md)
+
+## Proposal
+
+### High-level Requirement for Exposing Metrics
+
+Per-call stats from the perspective of the application should continue to be gathered from the interceptor/filter level, just as they are today.
+
+For each call, there should be some way to attach an object that will gather data for individual call attempts.  Events on the underlying call attempts will be recorded via that object.
+
+On the parent call object, there should be some way to record an event when we do the first LB pick attempt for a given call attempt.
+
+There should be some way to indicate whether a given call attempt was triggered via a transparent retry.
+
+With that general structure in the gRPC stats interface, it should be possible for whatever plugs into the stats interface (e.g., Census, OpenCensus, etc) to determine the data to export in the metrics defined here.
+
+
+### Metrics to Expose
+
+gRPC will treat each retry attempt or hedged RPC as a distinct RPC with regards to the current set of [metrics]((https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/gRPC.md#measures)).
```

Comment:
This link is also broken because of the parentheses.

Reply:
Done